core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stm
+
+/**
+ * Transactional map implemented on top of [[TRef]] and [[TArray]]. Resolves
+ * conflicts via chaining.
+ *
+ * Caution: doesn't provide stack-safety guarantees.
+ */
+class TMap[K, V] private (
+  private val tBuckets: TRef[TArray[List[(K, V)]]],
+  private val tCapacity: TRef[Int],
+  private val tSize: TRef[Int]
+) { self =>
+
+  /**
+   * Tests whether or not map contains a key.
+   */
+  final def contains[E](k: K): STM[E, Boolean] =
+    get(k).map(_.isDefined)
+
+  /**
+   * Removes binding for given key.
+   */
+  final def delete[E](k: K): STM[E, Unit] = {
```

Comment:
What is the reason for not just using Nothing?

Reply:
Good point. Applied the same change to `get`, `getOrElse` and `put`.